pkg/services/ngalert/notifier/file_store.go

```diff
@@ -0,0 +1,109 @@
+package notifier
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/grafana/grafana/pkg/infra/kvstore"
+)
+
+const KVNamespace = "alertmanager"
+
+// State represents any of the two 'states' of the alertmanager. Notification log or Silences.
+// MarshalBinary returns the binary representation of this internal state based on the protobuf.
+type State interface {
+	MarshalBinary() ([]byte, error)
+}
+
+// FileStore is in charge of persisting the alertmanager files to the database.
+// It uses the KVstore table and encodes the files as a base64 string.
+type FileStore struct {
+	kv             *kvstore.NamespacedKVStore
+	orgID          int64
+	workingDirPath string
+}
+
+func NewFileStore(orgID int64, store kvstore.KVStore, workingDirPath string) *FileStore {
+	return &FileStore{
+		workingDirPath: workingDirPath,
+		orgID:          orgID,
+		kv:             kvstore.WithNamespace(store, orgID, KVNamespace),
+	}
+}
+
+// FilepathFor returns the filepath to an Alertmanager file.
+// If the file is already present on disk it no-ops.
+// If not, it tries to read the database and if there's no file it no-ops.
+// If there is a file in the database, it decodes it and writes to disk for Alertmanager consumption.
+func (fs *FileStore) FilepathFor(filename string) (string, error) {
+	// If a file is already present, we'll use that one and eventually save it to the database.
+	// We don't need to do anything else.
+	if fs.IsExists(filename) {
+		return fs.pathFor(filename), nil
+	}
+
+	// Then, let's attempt to read it from the database.
+	content, exists, err := fs.kv.Get(context.Background(), filename)
```

Comment:
There's probably some merit in timing out on this call - not doing so poses the risk of sync taking too long. Doing so poses the risk of eventually overriding a configuration in the database...

Of the two, I chose the former until we get some customer feedback - I'm happy to be swayed the other way.